src/coreclr/inc/corinfo.h

```diff
@@ -1721,6 +1722,7 @@ struct CORINFO_FIELD_INFO
     CORINFO_HELPER_DESC     accessCalloutHelper;
 
     CORINFO_CONST_LOOKUP    fieldLookup;        // Used by Ready-to-Run
+    bool                    useEnclosingTypeOnly; // Used by NativeAOT
```

Comment:
What is the exact purpose of this flag? This needs a better comment.

Also, would it be better for it to be a new bit on existing flags instead of a new bool field?

Reply:
> What is the exact purpose of this flag?

Currently, we value number these helpers based on the `fieldLookup.addr` which is a symbol address that we obtain [here](https://github.com/dotnet/runtime/pull/89472/files#diff-756901b934d766cff71c53e0fa8c2f15fbc3cecdef7a392c8a2aacb2f6bc8a4eR2198-R2199) from tuple `<helperId, enclosingType>`. To eliminate the redundant check, whenever I see TLS access during importing, I am expanding `call CORINFO_HELP_READYTORUN_THREADSTATIC_BASE` into `(call CORINFO_HELP_READYTORUN_GCSTATIC_BASE, call CORINFO_HELP_READYTORUN_THREADSTATIC_BASE)`. However, the `fieldLookup.addr` for TLS access would be different than the one for `j` because the `helperId` used for creating the symbols were different. For TLS, it was `CORINFO_HELP_READYTORUN_THREADSTATIC_BASE` and for normal static variable, it was `CORINFO_HELP_READYTORUN_GCSTATIC_BASE`. With that, the VN will always be different for the helper call 
 `CORINFO_HELP_READYTORUN_GCSTATIC_BASE` *added* during TLS and that for `j`. To make sure that the *added* helper call gets same VN, we can just lookup based on the `enclosingType` and if they are same, we can have same VN for them, that will let us eventually eliminate the redundant call to `CORINFO_HELP_READYTORUN_GCSTATIC_BASE`.

Reply:
> Also, would it be better for it to be a new bit on existing flags instead of a new bool field?

I was thinking about that, but `CORINFO_FIELD_ACCESSOR` describes what type of access we do for the field and  `CORINFO_FIELD_FLAGS` tells if it is `static`, `protected`, etc., whereas `useEnclosingType` is just an information that we use for VN, so was not sure the right flag where I should add that information. Any preference on where it can go?

Reply:
perhaps, `CORINFO_FIELD_FLAGS` looks appropriate, but let me know if you disagree.

Reply:
that's true, but we don't consume the returned value from those helpers, so it works out well.

https://github.com/dotnet/runtime/blob/435c3238c3f3fa0dfe5c47a23812b3b869834b96/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/ClassConstructorRunner.cs#L33-L50

I could possibly add another field in `CORINFO_FIELD_INFO` to capture the correct symbol, but that might be unnecessary. @jkotas - any thoughts?

left:  ` = i; = j ;`
right: ` = j; = i;`


<img width="1673" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/bd6fd825-ccd8-4854-b4f3-284d0a4022cb">

Reply:
> I could possibly add another field in CORINFO_FIELD_INFO to capture the correct symbol, but that might be unnecessary

I think you can also adapt `getReadyToRunHelper` to get the entrypoint. In any case, I would be strongly against mis-appropriating entry points like this.

Reply:
no more needed.